designs/2023-only-run-reporting-rules/README.md

```diff
@@ -0,0 +1,126 @@
+- Repo: eslint/eslint
+- Start Date: 2023-01-16
+- RFC PR: <https://github.com/eslint/rfcs/pull/104>
+- Authors: Maddy Miller
+
+# Only run reporting lint rules
+
+## Summary
+
+Currently, ESLint will run all rules that are not marked as `off` in the configuration.
+This RFC proposes adding a way to configure which rules are actually run, to reduce linting
+time and better match the reporting outcome. Currently, when a rule is marked as `warn`,
+ESLint will still run the rule but not report the results when run under `--quiet`. The
+intended outcome of this RFC is to allow users to not run these rules when unnecessary.
+
+## Motivation
+
+A common use-case of the `warn` level is to report errors to the developer in their IDE,
+but to completely ignore it when running ESLint on the command line via the `--quiet` flag.
+When running with many rules set to the `warn` level, this can cause a significant amount
+of overhead on ESLint runtime. On an example large codebase, I found that by disabling all
+rules set to warn, I was able to reduce ESLint runtime from 4 minutes and 39 seconds to just
+22 seconds. As the `warn` rules are entirely ignored in this situation, this is a substantial
+amount of wasted time and resources that could be avoided.
+
+## Detailed Design
+
+The proposed implementation is to modify the `--quiet` flag so that ESLint will skip
+running any rules that are set to the `warn` level.
+
+From an API perspective, this would be implemented by a predicate function that acts as
```

Comment:
It seems to me that the main goal of the RFC is to fix the behavior of `--quiet`. So why is a new public Node API needed?

Reply:
The addition of API comes from comments on the initial issue before this RFC, https://github.com/eslint/eslint/issues/16450#issuecomment-1286780419

Reply:
As far as I can tell, tweaking the behavior of `--quiet` already addresses the problem that triggered the issue/RFC, so I don't think an API for filtering out rules is necessary anymore. If it was necessary for some reason, the RFC would need to spell out the motivation and use cases for it as well as an example of what it's usage looks like. CC: @mdjermanovic.

Reply:
So to clarify, if the API needs to change just for the `--quiet` flag to make the requested change, it doesn't need to be included in the RFC? If so, I'm happy to remove the mentions of it. As you said, my motivation for this RFC is purely to improve behaviour of the `--quiet` flag.

Reply:
I'm a bit confused by your question. But I agree with this:

> my motivation for this RFC is purely to improve behaviour of the `--quiet` flag

Reply:
> It seems to me that the main goal of the RFC is to fix the behavior of `--quiet`. So why is a new public Node API needed?

`--quiet` mode currently exists at the CLI level only. CLI uses API (ESLint/FlatESLint class), which is not aware of the quiet mode and just returns all errors and warnings as usual. Then, CLI [filters out](https://github.com/eslint/eslint/blob/2aa4f5fb2fdb1c4a1734093c225e5c6251b0ee0f/lib/cli.js#L427-L430) warnings to pass only errors to the formatter. If we want to skip running rules set to warn when the `--quiet` CLI flag is used, I believe that change has to be implemented in FlatESLint and/or Linter, which are API-level modules. So we'll need a new API option that will change the way it operates, e.g., boolean `quiet` or something generic like a predicate function.

Reply:
Oh, got it, that makes sense now.

Reply:
I've just pushed a change that goes into more detail here. I'm not extremely familiar with the internals of ESLint, but I believe what I've described in the updated document is correct - but if anyone has a better idea please let me know

Reply:
Thanks for the additional details. Part of the RFC is designed to get you familiar with ESLint internals. The goal is to produce an RFC that someone else has a reasonable chance of implementing, much like a tech spec.

To that end, while your description makes sense at a high level, we really need more low level details. Where exactly will this predicate function exist? What API will it be passed into?

To start, please look at `lib/cli.js`. As @mdjermanovic suggested earlier, it is most likely that something would have to be passed to the `ESLint` constructor, which also means it needs to be implemented in the `FlatESLint` constructor.

Reply:
I've filled out the section specifying specific function calls and where it would be added. I've validated this by writing a patch that implements this following the instructions set out in the RFC

Reply:
Awesome work @me4502!

Reply:
Since this is a new public API, can you include a code sample demonstrating how someone would use the API?

And can you include a link to your patch here or in the PR description in case we want to take a look?

Reply:
I've just updated the description to include my rough patch file, and the RFC to include an example usage of the API implementing this change.